// lighter-type/lighter-type.js
'use strict'
/**
 * A Type can be instantiated and extended to yield sub-types. whose instances can be constructed
 * with an `init` method.
 */

// The base constructor does nothing.
var Type = module.exports = function Type () {}

/**
 * Extend a Type, yielding a new Type with additional properties from a map.
 *
 * @param  {Object} pros  An optional map of additional prototype properties.
 * @param  {Object} typs  An optional map of additional type properties.
 * @return {Object}      The new Type.
 */
Type.extend = function extend (pros, typs) {
  // Construct.
  var type = (pros || 0).init || function SubType () {
    type._super.apply(this, arguments)
  }
  var pro = type.prototype

  // Copy.
  this.decorate(type, this, true)
  this.decorate(pro, this.prototype, true)

  // Extend.
  if (pros) {
    for (var key in pros) {
      if (key !== 'init') {
        pro[key] = pros[key]
      }
    }
  }
  if (typs) {
    this.decorate(type, typs, true)
  }

  // Link.
  this.hide(type, '_super', this)

  return type
}

/**
 * Decorate an object with prototype properties, and run a constructor on it.
 *
 * @param  {Object}         object     An object to decorate.
 * @param  {Boolean}        overwrite  Whether to overwrite existing properties.
 * @param  {Array|Boolean}  args       Optional arguments for the constructor,
 *                                     or false to skip the constructor.
 */
Type.init = function init (object, overwrite, args) {
  // Allow calling with (object, args).
  if (overwrite && overwrite.length) {
    args = overwrite
    overwrite = false
  }
  this.decorate(object, this.prototype, overwrite)
  if (args !== false) {
    this.apply(object, args)
  }
}

/**
 * Decorate an object with a map of properties.
 *
 * @param  {Object}  object     An object to decorate.
 * @param  {Object}  map        An optional map to decorate the object with.
 * @param  {Boolean} overwrite  Whether to overwrite existing properties.
 */
Type.decorate = function decorate (object, map, overwrite) {
  for (var key in map) {
    if (overwrite || (object[key] === undefined)) {
      object[key] = map[key]
    }
  }
}

/**
 * Include another type's prototype methods in this one's.
 *
 * @param  {Object}  type       Another type to mix in.
 * @param  {Boolean} overwrite  Whether to overwrite existing properties.
 */
Type.include = function include (type, overwrite) {
  this.decorate(this.prototype, type.prototype, overwrite)
  var includes = this._includes
  if (includes) {
    includes[includes.length] = type
  } else {
    Type.hide(this, '_includes', [type])
  }
}

/**
 * Define a non-enumerable property on an object.
 *
 * @param  {Object} object  An object to define a property on.
 * @param  {String} key     The name of the property to define.
 * @param  {Any}    value   The value of the property to define.
 */
Type.hide = function hide (object, key, value) {
  Object.defineProperty(object, key, {
    enumerable: false,
    writable: true,
    value: value
  })
}

/**
 * Checks whether this Type is an extension of another Type.
 *
 * @param  {Type}    type  A possible ancestor.
 * @return {Boolean}       True if this type is an extension of the given type.
 */
Type.is = function is (type) {
  var sup = this
  while (sup) {
    if (sup === type) {
      return true
    }
    sup = sup._super
  }
  return false
}

/**
 * Checks whether this Type has acquired the functionality of another type
 * via the extend method or the include method.
 *
 * @param  {Type}    type  A type whose functionality might be mixed in.
 * @return {Boolean}       True the given type is extended from or mixed in.
 */
Type.has = function has (type) {
  if (this === type) {
    return true
  }
  var sup = this._super
  if (sup) {
    if (sup === type) {
      return true
    }
    if (sup.has(type)) {
      return true
    }
  }
  var includes = this._includes
  if (includes) {
    for (var i = 0, l = includes.length; i < l; i++) {
      if (includes[i].has(type)) {
        return true
      }
    }
  }
  return false
}
